<!DOCTYPE html PUBLIC "-//W3C//DTD HTML 4.01//EN" "http://www.w3.org/TR/html4/strict.dtd">
<html>
<head>
  <meta http-equiv="Content-Type" content="text/html; charset=utf-8">
  <meta http-equiv="Content-Style-Type" content="text/css">
  <title></title>
  <meta name="Generator" content="Cocoa HTML Writer">
  <meta name="CocoaVersion" content="2299.77">
  <style type="text/css">
    p.p1 {margin: 6.0px 0.0px 3.0px 0.0px; font: 11.0px Optima}
    p.p2 {margin: 2.0px 0.0px 2.0px 0.0px; text-indent: 13.7px; font: 11.0px Optima; color: #000000}
    p.p3 {margin: 9.0px 0.0px 9.0px 27.4px; font: 9.0px Menlo; color: #000000}
    p.p4 {margin: 0.0px 0.0px 3.0px 94.5px; text-indent: -67.2px; font: 11.0px Optima; color: #000000}
    p.p5 {margin: 0.0px 0.0px 9.0px 94.5px; text-indent: -67.2px; font: 11.0px Optima; color: #000000}
    p.p6 {margin: 9.0px 0.0px 9.0px 27.4px; font: 11.0px Optima; color: #000000}
    p.p7 {margin: 2.0px 0.0px 2.0px 0.0px; font: 11.0px Optima; color: #000000}
    p.p8 {margin: 0.0px 0.0px 0.0px 94.3px; text-indent: -67.0px; font: 11.0px Optima; color: #000000}
    p.p9 {margin: 0.0px 0.0px 0.0px 103.7px; text-indent: -76.3px; font: 11.0px Optima; color: #000000}
    p.p10 {margin: 0.0px 0.0px 9.0px 103.5px; text-indent: -76.2px; font: 11.0px Optima; color: #000000}
    p.p11 {margin: 0.0px 0.0px 0.0px 54.0px; text-indent: -26.6px; font: 11.0px Optima; color: #000000}
    p.p12 {margin: 0.0px 0.0px 9.0px 54.0px; text-indent: -26.6px; font: 11.0px Optima; color: #000000}
    p.p13 {margin: 0.0px 0.0px 9.0px 103.7px; text-indent: -76.3px; font: 11.0px Optima; color: #000000}
    p.p14 {margin: 2.0px 0.0px 2.0px 0.0px; text-indent: 13.7px; font: 11.0px Optima; color: #000000; min-height: 13.0px}
    span.s1 {font: 9.0px Menlo}
    span.s2 {color: #1c00cf}
    span.s3 {color: #3f6e74}
    span.s4 {color: #1400c4}
    span.s5 {color: #b50013}
    span.s6 {font: 11.0px Optima}
    span.s7 {font: 7.3px Optima}
    span.s8 {font-kerning: none}
    span.Apple-tab-span {white-space:pre}
  </style>
</head>
<body>
<p class="p1"><i>5.13.0<span class="Apple-converted-space">  </span>ITEM: 1. </i><span class="s1"><i>initialize()</i></span><i> callbacks</i></p>
<p class="p2">Before a SLiM simulation can be run, the various classes underlying the simulation need to be set up with an initial configuration.<span class="Apple-converted-space">  </span>Simulation configuration in SLiM is done in <span class="s1"><i>initialize()</i></span><i> callbacks</i> that run prior to the beginning of simulation execution.<span class="Apple-converted-space">  </span>For our present purposes, the idea is very simple; in your input file, you can write something like this:</p>
<p class="p3">initialize()<br>
{<br>
<span class="Apple-tab-span">	</span>...<br>
}</p>
<p class="p2">The <span class="s1">initialize()</span> declaration specifies that the script block is to be executed as an <span class="s1">initialize()</span> callback before the simulation starts.<span class="Apple-converted-space">  </span>The script between the braces <span class="s1">{}</span> would set up various aspects of the simulation by calling <i>initialization functions</i>.<span class="Apple-converted-space">  </span>These are SLiM functions that may be called only in an <span class="s1">initialize()</span> callback, and their names begin with <span class="s1">initialize</span> to mark them clearly as such.<span class="Apple-converted-space">  </span>You may also use other Eidos functionality in these callbacks; for example, you might automate generating a complex genetic structure containing many genes by using a <span class="s1">for</span> loop.</p>
<p class="p2">In general, it is required for a species to set up its genetic structure in an <span class="s1">initialize()</span> callback with calls to <span class="s1">initializeMutationRate()</span>, <span class="s1">initializeRecombinationRate()</span>, <span class="s1">initializeMutationType()</span>, <span class="s1">initializeGenomicElementType()</span>, and <span class="s1">initializeGenomicElement()</span>; species must call all of these, setting up at least one mutation type, at least one genomic element type, and at least one genomic element.<span class="Apple-converted-space">  </span>The exception to this general rule is for species that have no genetics at all – species that are modeled purely on an ecological/behavioral level.<span class="Apple-converted-space">  </span>Such species may be defined by calling <i>none</i> of those initialization functions; in this case, SLiM will default to a zero-length chromosome with mutation and recombination rates of zero.<span class="Apple-converted-space">  </span>A middle ground between these two configuration paths is not allowed; either a species has no genetics, or it fully defines its genetics.</p>
<p class="p2">One thing worth mentioning is that in the context of an <span class="s1">initialize()</span> callback, the <span class="s1">sim</span> global representing the species being simulated is not defined.<span class="Apple-converted-space">  </span>This is because the state of the simulation is not yet constructed fully, and accessing partially constructed state would not be safe.<span class="Apple-converted-space">  </span>(Similarly, in multispecies models, the <span class="s1">community</span> object and the objects representing individual species are not yet defined.)</p>
<p class="p2">The above <span class="s1">initialize()</span> callback syntax <i>implicitly</i> declares a single species, with the default name of <span class="s1">sim</span>, and therefore sets up a single-species model.<span class="Apple-converted-space">  </span>It is also possible to <i>explicitly</i> declare a species, which is done with this extended syntax (using a species name of <span class="s1">fox</span> as an example):</p>
<p class="p3">species fox initialize() { ... }</p>
<p class="p2">This sets up a multispecies model (although it might, in fact, declare only a single species, <span class="s1">fox</span>; the term “multispecies”, in SLiM parlance, really means “explicitly declared species”, but multispecies models almost always <i>do</i> contain multiple species, so the distinction is unimportant).<span class="Apple-converted-space">  </span>In most respects multispecies models work identically to single-species models, so we will tend to focus on the single-species case in the reference documentation, with a species name of <span class="s1">sim</span>, for simplicity and clarity.</p>
<p class="p2">In single-species models all initialization can be done in a single <span class="s1">initialize()</span> callback (or you can have more than one, if you wish).<span class="Apple-converted-space">  </span>In multispecies models, each species must be initialized with its own callback(s), as shown above.<span class="Apple-converted-space">  </span>In addition, multispecies models also support an optional community-level initialization callback that is declared as follows:</p>
<p class="p3">species all initialize() { ... }</p>
<p class="p2">These callbacks, technically called <i>non-species-specific </i><span class="s1"><i>initialize()</i></span><i> callbacks</i>, provide a place for community-level initialization to occur.<span class="Apple-converted-space">  </span>They are run before any species-specific <span class="s1">initialize()</span> callbacks are run, so you might wish to set up all of your model parameters in one, providing a single location for all parameters.<span class="Apple-converted-space">  </span>In multispecies models, the <span class="s1">initializeModelType()</span> and <span class="s1">initializeInteractionType()</span> functions may only be called from a non-species-specific <span class="s1">initialize()</span> callback, since those aspects of model configuration span the entire community.<span class="Apple-converted-space">  </span>In single-species models, these functions may be called from an ordinary <span class="s1">initialize()</span> callback for simplicity and backward compatibility.</p>
<p class="p2">Once all <span class="s1">initialize()</span> callbacks have executed, in the order in which they are specified in the SLiM input file, the simulation will begin.<span class="Apple-converted-space">  </span>The tick number at which it starts is determined by the Eidos events you have defined; the first tick in which an Eidos event is scheduled to execute is the tick at which the simulation starts.<span class="Apple-converted-space">  </span>Similarly, the simulation will terminate after the last tick for which a script block (either an event or a callback) is registered to execute, unless the <span class="s1">stop()</span> function or the <span class="s1">simulationFinished()</span> method of <span class="s1">Community</span> or <span class="s1">Species</span> are called to end the simulation earlier.</p>
<p class="p1"><i>5.13.1<span class="Apple-converted-space">  </span>ITEM: 2. Eidos events</i></p>
<p class="p2">An Eidos event is a block of Eidos code that is executed every tick, within a tick range, to perform a desired task.<span class="Apple-converted-space">  </span>The syntax of an Eidos event declaration looks like one of these:</p>
<p class="p3">[id] [t1 [: t2]] first() { ... }<br>
[id] [t1 [: t2]] early() { ... }<br>
[id] [t1 [: t2]] late() { ... }</p>
<p class="p2">The first declaration declares a <span class="s1">first()</span> event that executes first thing in the tick cycle.<span class="Apple-converted-space">  </span>The second declaration declares an <span class="s1">early()</span> event that executes relatively early in the tick cycle.<span class="Apple-converted-space">  </span>The third declaration declares a <span class="s1">late()</span> event that executes near the end of the tick cycle.<span class="Apple-converted-space">  </span>Exactly when these events run depends upon whether the model is a WF model or a nonWF model; see the tick cycle diagrams for those model types.</p>
<p class="p2">The <span class="s1">id</span> is an optional identifier like <span class="s1">s1</span> (or more generally, <span class="s1">sX</span>, where <span class="s1">X</span> is an integer greater than or equal to <span class="s1">0</span>) that defines an identifier that can be used to refer to the script block.<span class="Apple-converted-space">  </span>In most situations it can be omitted, in which case the <span class="s1">id</span> is implicitly defined as <span class="s1">-1</span>, a placeholder value that essentially represents the lack of an identifier value.<span class="Apple-converted-space">  </span>Supplying an <span class="s1">id</span> is only useful if you wish to manipulate your script blocks programmatically.</p>
<p class="p2">Then comes a tick or a range of ticks, and then a block of Eidos code enclosed in braces to form a compound statement.<span class="Apple-converted-space">  </span>A trivial example might look like this:</p>
<p class="p3"><span class="s2">1000</span>:<span class="s2">5000</span> early() {<br>
<span class="Apple-tab-span">	</span>catn(<span class="s3">community</span>.tick);<br>
}</p>
<p class="p2">This would print the tick number in every tick in the specified range, which is obviously not very exciting.<span class="Apple-converted-space">  </span>The broader point is that the Eidos code in the braces <span class="s1">{}</span> is executed early in every tick within the specified range of ticks.<span class="Apple-converted-space">  </span>In this case, the tick range is <span class="s1">1000</span> to <span class="s1">5000</span>, and so the Eidos event will be executed 4001 times (not 4000!).<span class="Apple-converted-space">  </span>A range of ticks can be given, as in the example above, or a single tick can be given with a single integer:</p>
<p class="p3"><span class="s4">100</span> late() {<br>
<span class="Apple-tab-span">	</span>print(<span class="s5">"Finished tick 100!"</span>);<br>
}</p>
<p class="p2">The tick range may also be incompletely specified, with a somewhat idiosyncratic syntax.<span class="Apple-converted-space">  </span>A range of <span class="s1">1000:</span> would specify that the event should run in tick <span class="s1">1000</span> and every subsequent tick until the model finishes; a range of <span class="s1">:1000</span> would similarly specify that the event should run in the first tick executed, and every subsequent tick, up to and including tick <span class="s1">1000</span>.</p>
<p class="p2">In fact, you can omit specifying a tick altogether, in which case the Eidos event runs every tick.<span class="Apple-converted-space">  </span>Since it takes a little time to set up the Eidos interpreter and interpret a script, it is advisable to use the narrowest range of ticks possible; however, that is more of a concern with callbacks, since they might be called many time in every tick, whereas <span class="s1">first()</span>, <span class="s1">early()</span>, and <span class="s1">late()</span> events will just be called once per tick.</p>
<p class="p2">The ticks specified for a Eidos event block can be any positive integer.<span class="Apple-converted-space">  </span>All blocks that apply to a given time point will be run in <i>definition order</i>; blocks specified higher in the input file will run before those specified lower.<span class="Apple-converted-space">  </span>Sometimes it is desirable to have a script block execute in a tick which is not fixed, but instead depends upon some parameter, defined constant, or calculation; this may be achieved by rescheduling the script block with the <span class="s1">Community</span> method <span class="s1">rescheduleScriptBlock()</span>.</p>
<p class="p2">In multispecies models, one can optionally provide a <span class="s1">ticks</span> specifier before the definition of an Eidos event, specifying that the event should only run in ticks in which a particular species is active.<span class="Apple-converted-space">  </span>That extended syntax looks like this:</p>
<p class="p3">[ticks species_name] [id] [t1 [: t2]] first() { ... }<br>
[ticks species_name] [id] [t1 [: t2]] early() { ... }<br>
[ticks species_name] [id] [t1 [: t2]] late() { ... }</p>
<p class="p2">The <span class="s1">species_name</span> should be the name of a species that was explicitly declared in the multispecies model.<span class="Apple-converted-space">  </span>If the <span class="s1">ticks</span> specifier is omitted, the event will run in every tick (within the specified tick range).</p>
<p class="p2">When Eidos events are executed, several global variables are defined by SLiM for use by the Eidos code.<span class="Apple-converted-space">  </span>Here is a summary of those SLiM globals:</p>
<p class="p4"><span class="s1">community</span><span class="Apple-tab-span">	</span>The <span class="s1">Community</span> object for the overall simulation</p>
<p class="p4"><span class="s1">sim</span><span class="Apple-tab-span">	</span>A <span class="s1">Species</span> object for the simulated species (in single-species simulations)</p>
<p class="p4"><span class="s1">g1</span>, <span class="s1">...</span><span class="Apple-tab-span">	</span><span class="s1">GenomicElementType</span> objects for defined genomic element types</p>
<p class="p4"><span class="s1">i1</span>, <span class="s1">...</span><span class="Apple-tab-span">	</span><span class="s1">InteractionType</span> objects for defined interaction types</p>
<p class="p4"><span class="s1">m1</span>, <span class="s1">...</span><span class="Apple-tab-span">	</span><span class="s1">MutationType</span> objects representing defined mutation types</p>
<p class="p4"><span class="s1">p1</span>, <span class="s1">...</span><span class="Apple-tab-span">	</span><span class="s1">Subpopulation</span> objects for existing subpopulations</p>
<p class="p4"><span class="s1">s1</span>, <span class="s1">...</span><span class="Apple-tab-span">	</span><span class="s1">SLiMEidosBlock</span> objects for named events and callbacks</p>
<p class="p5"><span class="s1">self</span><span class="Apple-tab-span">	</span>A <span class="s1">SLiMEidosBlock</span> object for the script block currently executing</p>
<p class="p2">In multispecies models, symbols for each species will be defined instead of <span class="s1">sim</span>.<span class="Apple-converted-space">  </span>Note that species symbols such as <span class="s1">sim</span> are <i>not</i> available in <span class="s1">initialize()</span> callbacks, since the species objects have not yet been initialized.<span class="Apple-converted-space">  </span>Similarly, the globals for subpopulations, mutation types, and genomic element types are only available after the point at which those objects have been defined by an <span class="s1">initialize()</span> callback.</p>
<p class="p1"><i>5.13.2<span class="Apple-converted-space">  </span>ITEM: 3. </i><span class="s1"><i>mutationEffect()</i></span><i> callbacks</i></p>
<p class="p2">An Eidos callback is a block of Eidos code that is called by SLiM in specific circumstances, to allow the customization of particular actions taken by SLiM while running the simulation of a species.<span class="Apple-converted-space">  </span>Nine types of callbacks are presently supported (in addition to <span class="s1">initialize()</span> callbacks): <span class="s1">mutationEffect()</span> callbacks, discussed here, and <span class="s1">fitnessEffect()</span>, <span class="s1">mateChoice()</span>, <span class="s1">modifyChild()</span>, <span class="s1">recombination()</span>, <span class="s1">interaction()</span>, <span class="s1">reproduction()</span> , <span class="s1">mutation()</span>, and <span class="s1">survival()</span> callbacks.</p>
<p class="p2">A <span class="s1">mutationEffect()</span> callback is called by SLiM when it is determining the fitness effect of a mutation carried by an individual.<span class="Apple-converted-space">  </span>Normally, the fitness effect of a mutation is determined by the selection coefficient <i>s</i> of the mutation and the dominance coefficient <i>h</i> of the mutation (the latter used only if the individual is heterozygous for the mutation).<span class="Apple-converted-space">  </span>More specifically, the standard calculation for the fitness effect of a mutation takes one of two forms.<span class="Apple-converted-space">  </span>If the individual is homozygous, then the fitness effect is (1+<i>s</i>), or:</p>
<p class="p6"><i>w</i> = <i>w</i> * (1.0 + selectionCoefficient),</p>
<p class="p7">where <i>w</i> is the relative fitness of the individual carrying the mutation.<span class="Apple-converted-space">  </span>If the individual is heterozygous, then the dominance coefficient enters the picture, and the fitness effect is (1+<i>hs</i>) or:</p>
<p class="p6"><i>w</i> = <i>w</i> * (1.0 + dominanceCoeff * selectionCoeff).</p>
<p class="p2">The dominance coefficient usually comes from the <span class="s1">dominanceCoeff</span> property of the mutation’s <span class="s1">MutationType</span>; if the focal individual has only one non-null haplosome, however, such that the mutation is paired with a null haplosome (i.e., is actually hemizygous, not heterozygous), the <span class="s1">hemizygousDominanceCoeff</span> property of the <span class="s1">MutationType</span> is used instead.</p>
<p class="p2">That is the standard behavior of SLiM, reviewed here to provide a conceptual baseline.<span class="Apple-converted-space">  </span>Supplying a <span class="s1">mutationEffect()</span> callback allows you to substitute any calculation you wish for the relative fitness effect of a mutation; the new relative fitness effect computation becomes:</p>
<p class="p3"><span class="s6"><i>w</i> = <i>w</i> * </span>mutationEffect()</p>
<p class="p7">where <span class="s1">mutationEffect()</span> is the value returned by your callback.<span class="Apple-converted-space">  </span>This value is a multiplicative fitness effect, so <span class="s1">1.0</span> is neutral, unlike the selection coefficient scale where <span class="s1">0.0</span> is neutral; be careful with this distinction!</p>
<p class="p2">Like Eidos events, <span class="s1">mutationEffect()</span> callbacks are defined as script blocks in the input file, but they use a variation of the syntax for defining an Eidos event:</p>
<p class="p3">[id] [t1 [: t2]] mutationEffect(&lt;mut-type-id&gt; [, &lt;subpop-id&gt;]) { ... }</p>
<p class="p2">For example, if the callback were defined as:</p>
<p class="p3">1000:2000 mutationEffect(m2, p3) { 1.0; }</p>
<p class="p7">then a relative fitness of <span class="s1">1.0</span> (i.e. neutral) would be used for all mutations of mutation type <span class="s1">m2</span> in subpopulation <span class="s1">p3</span> from tick <span class="s1">1000</span> to tick <span class="s1">2000</span>.<span class="Apple-converted-space">  </span>The very same mutations, if also present in individuals in other subpopulations, would preserve their normal selection coefficient and dominance coefficient in those other subpopulations; this callback would therefore establish spatial heterogeneity in selection, in which mutation type <span class="s1">m2</span> was neutral in subpopulation <span class="s1">p3</span> but under selection in other subpopulations, for the range of ticks given.</p>
<p class="p2">In multispecies models, callbacks must be defined with a <span class="s1">species</span> specifier that states the species with which species the callback is associated.<span class="Apple-converted-space">  </span>Such a definition looks like this:</p>
<p class="p3">species species_name [id] [t1 [: t2]] mutationEffect(...) { ... }</p>
<p class="p2">It is the same syntax, in other words, except for the <span class="s1">species</span> specifier at the beginning, with the name of the species that the callback will modify.<span class="Apple-converted-space">  </span>As with the <span class="s1">ticks</span> specifier for events, this means the callback will only be called in ticks when the species is active; but the <span class="s1">species</span> specifier goes further, making that species the focal species for the callback.</p>
<p class="p2">In addition to the standard SLiM globals, a <span class="s1">mutationEffect()</span> callback is supplied with some additional information passed through “pseudo-parameters”, variables that are defined by SLiM within the context of the callback’s code to supply the callback with relevant information:</p>
<p class="p8"><span class="s1">mut</span><span class="Apple-tab-span">	</span>A <span class="s1">Mutation</span> object, the mutation whose relative fitness is being evaluated</p>
<p class="p8"><span class="s1">homozygous</span><span class="Apple-tab-span">	</span>A value of <span class="s1">T</span> (the mutation is homozygous), <span class="s1">F</span> (heterozygous), or <span class="s1">NULL</span> (it is<br>
paired with a null haplosome, and is thus hemizygous or haploid)</p>
<p class="p8"><span class="s1">effect</span><span class="Apple-tab-span">	</span>The default relative fitness value calculated by SLiM</p>
<p class="p8"><span class="s1">individual</span><span class="Apple-tab-span">	</span>The individual carrying this mutation (an object of class <span class="s1">Individual</span>)</p>
<p class="p5"><span class="s1">subpop</span><span class="Apple-tab-span">	</span>The subpopulation in which that individual lives</p>
<p class="p2">These may be used in the <span class="s1">mutationEffect()</span> callback to compute a fitness value.<span class="Apple-converted-space">  </span>To implement the standard fitness functions used by SLiM for an autosomal simulation with no null haplosomes involved, for example, you could do something like this:</p>
<p class="p3">mutationEffect(m1) {<br>
<span class="Apple-tab-span">	</span>if (homozygous)<br>
<span class="Apple-tab-span">	</span><span class="Apple-tab-span">	</span>return 1.0 + mut.selectionCoeff;<br>
<span class="Apple-tab-span">	</span>else<br>
<span class="Apple-tab-span">	</span><span class="Apple-tab-span">	</span>return 1.0 + mut.mutationType.dominanceCoeff * mut.selectionCoeff;<br>
}</p>
<p class="p2">As mentioned above, a relative fitness of <span class="s1">1.0</span> is neutral (whereas a selection coefficient of <span class="s1">0.0</span> is neutral); the <span class="s1">1.0 +</span> in these calculations converts between the selection coefficient scale and the relative fitness scale, and is therefore essential.<span class="Apple-converted-space">  </span>However, the <span class="s1">effect</span> global variable mentioned above would already contain this value, precomputed by SLiM, so you could simply return <span class="s1">effect</span> to get that behavior when you want it:</p>
<p class="p3">mutationEffect(m1) {<br>
<span class="Apple-tab-span">	</span>if (&lt;conditions&gt;)<br>
<span class="Apple-tab-span">	</span><span class="Apple-tab-span">	</span>&lt;custom fitness calculations...&gt;;<br>
<span class="Apple-tab-span">	</span>else<br>
<span class="Apple-tab-span">	</span><span class="Apple-tab-span">	</span>return effect;<br>
}</p>
<p class="p2">This would return a modified fitness value in certain conditions, but would return the standard fitness value otherwise.</p>
<p class="p2">More than one <span class="s1">mutationEffect()</span> callback may be defined to operate in the same tick.<span class="Apple-converted-space">  </span>As with Eidos events, multiple callbacks will be called in the order in which they were defined in the input file.<span class="Apple-converted-space">  </span>Furthermore, each callback will be given the <span class="s1">effect</span> value returned by the previous callback – so the value of <span class="s1">effect</span> is not necessarily the default value, in fact, but is the result of all previous <span class="s1">mutationEffect()</span> callbacks for that individual in that tick.<span class="Apple-converted-space">  </span>In this way, the effects of multiple callbacks can “stack”.</p>
<p class="p2">One caveat to be aware of in WF models is that <span class="s1">mutationEffect()</span> callbacks are called at the end of the tick, just before the next tick begins.<span class="Apple-converted-space">  </span>If you have a <span class="s1">mutationEffect()</span> callback defined for tick <span class="s1">10</span>, for example, it will actually be called at the very end of tick <span class="s1">10</span>, after child generation has finished, after the new children have been promoted to be the next parental generation, and after <span class="s1">late()</span> events have been executed.<span class="Apple-converted-space">  </span>The fitness values calculated will thus be used during tick <span class="s1">11</span>; the fitness values used in tick <span class="s1">10</span> were calculated at the end of tick <span class="s1">9</span>.<span class="Apple-converted-space">  </span>(This is primarily so that SLiMgui, which refreshes its display in between ticks, has computed fitness values at hand that it can use to display the new parental individuals in the proper colors.)<span class="Apple-converted-space">  </span>This is not an issue in nonWF models, since fitness values are used in the same tick in which they are calculated.</p>
<p class="p2">If the <span class="s1">randomizeCallbacks</span> parameter to <span class="s1">initializeSLiMOptions()</span> is <span class="s1">T</span> (the default), the order in which the fitness of individuals is evaluated will be randomized within each subpopulation.<span class="Apple-converted-space">  </span>This partially mitigates order-dependency issues, although such issues can still arise whenever the effects of a <span class="s1">mutationEffect()</span> callback are not independent.<span class="Apple-converted-space">  </span>If <span class="s1">randomizeCallbacks</span> is <span class="s1">F</span>, the fitness of individuals will be evaluated in sequential order within each subpopulation, greatly increasing the risk of order-dependency problems.</p>
<p class="p2">Many other possibilities can be implemented with <span class="s1">mutationEffect()</span> callbacks.<span class="Apple-converted-space">  </span>However, since <span class="s1">mutationEffect()</span> callbacks involve Eidos code being executed for the evaluation of fitness of every mutation of every individual (within the tick range, mutation type, and subpopulation specified), they can slow down a simulation considerably, so use them as sparingly as possible.</p>
<p class="p1"><i>5.13.3<span class="Apple-converted-space">  </span>ITEM: 4. </i><span class="s1"><i>fitnessEffect()</i></span><i> callbacks</i></p>
<p class="p2">We have already seen <span class="s1">mutationEffect()</span> callbacks, which modify the effect of a given mutation in a focal individual.<span class="Apple-converted-space">  </span>Sometimes it is desirable to model effects upon individual fitness that are not governed by particular mutations (or not directly, at least); fitness effects due to spatial position, or resource acquisition, or behavior such as competitive or altruistic interactions, for example.<span class="Apple-converted-space">  </span>Another situation of this type is when fitness depends upon the overall phenotype of an individual – the height of a tree, say – which might be influenced by genetics, but also by environmental effects, climate, and so forth.<span class="Apple-converted-space">  </span>For these sorts of situations, SLiM provides <span class="s1">fitnessEffect()</span> callbacks.</p>
<p class="p2">A <span class="s1">fitnessEffect()</span> callback is called by SLiM when it is determining the fitness of an individual – typically, but not always, once per tick during the fitness calculation tick cycle stage.<span class="Apple-converted-space">  </span>Normally, the fitness of a given individual is determined by multiplying together the fitness effects of all mutations possessed by that individual.<span class="Apple-converted-space">  </span>Supplying a <span class="s1">fitnessEffect()</span> callback allows you to add another multiplicative fitness effect into that calculation.<span class="Apple-converted-space">  </span>As with <span class="s1">mutationEffect()</span> callbacks, the value returned by <span class="s1">fitnessEffect()</span> callbacks is a fitness effect, so <span class="s1">1.0</span> is neutral.</p>
<p class="p2">The syntax for declaring <span class="s1">fitnessEffect()</span> callbacks is similar to that for <span class="s1">mutationEffect()</span> callbacks, but simpler since no mutation type is needed:</p>
<p class="p3">[id] [t1 [: t2]] fitnessEffect([&lt;subpop-id&gt;]) { ... }</p>
<p class="p2">(In multispecies models, the definition must be preceded by a <span class="s1">species</span> specification as usual.)</p>
<p class="p2">For example, if the callback were defined as:</p>
<p class="p3">1000:2000 fitnessEffect(p3) { 0.75; }</p>
<p class="p2">then a fitness effect of <span class="s1">0.75</span> would be multiplied into the fitness values of all individuals in subpopulation <span class="s1">p3</span> from tick <span class="s1">1000</span> to tick <span class="s1">2000</span>.</p>
<p class="p2">Much more interesting, of course, are <span class="s1">fitnessEffect()</span> callbacks that return different fitness effects for different individuals, depending upon their state!<span class="Apple-converted-space">  </span>In addition to the standard SLiM globals, a <span class="s1">fitnessEffect()</span> callback is supplied with some additional information passed through “pseudo-parameters”, variables that are defined by SLiM within the context of the callback’s code to supply the callback with relevant information:</p>
<p class="p8"><span class="s1">individual</span><span class="Apple-tab-span">	</span>The focal individual (an object of class <span class="s1">Individual</span>)</p>
<p class="p5"><span class="s1">subpop</span><span class="Apple-tab-span">	</span>The subpopulation in which that individual lives</p>
<p class="p2">These may be used in the <span class="s1">fitnessEffect()</span> callback to compute a fitness effect that depends upon the state of the focal individual.<span class="Apple-converted-space">  </span>The fitness effect for the callback is simply returned as a singleton <span class="s1">float</span> value, as usual.</p>
<p class="p2">More than one <span class="s1">fitnessEffect()</span> callback may be defined to operate in the same tick.<span class="Apple-converted-space">  </span>Each such callback will provide an independent fitness effect for the focal individual; the results of each <span class="s1">fitnessEffect()</span> callback will be multiplied in to the individual’s fitness.<span class="Apple-converted-space">  </span>These callbacks will generally be called once per individual in each tick, in an order that is formally undefined.</p>
<p class="p2">Beginning in SLiM 3.0, it is also possible to set the <span class="s1">fitnessScaling</span> property on a subpopulation to scale the fitness values of every individual in the subpopulation by the same constant amount, or to set the <span class="s1">fitnessScaling</span> property on an individual to scale the fitness value of that specific individual.<span class="Apple-converted-space">  </span>These scaling factors are multiplied together with all other fitness effects for an individual to produce the individual’s final fitness value.<span class="Apple-converted-space">  </span>The <span class="s1">fitnessScaling</span> properties of <span class="s1">Subpopulation</span> and <span class="s1">Individual</span> can often provide similar functionality to <span class="s1">fitnessEffect()</span> callbacks with greater efficiency and simplicity.<span class="Apple-converted-space">  </span>They are reset to <span class="s1">1.0</span> in every tick for which a given species is active, immediately after fitness values are calculated, so they only need to be set when a value other than <span class="s1">1.0</span> is desired.</p>
<p class="p2">As with <span class="s1">mutationEffect()</span> callbacks, <span class="s1">fitnessEffect()</span> callbacks are called at the end of the tick, just before the next tick begins.<span class="Apple-converted-space">  </span>Also, as with <span class="s1">mutationEffect()</span> callbacks, the order in which <span class="s1">fitnessEffect()</span> callbacks are called will be shuffled when <span class="s1">randomizeCallbacks</span> is enabled, as it is by default, partially mitigating order-dependency issues.</p>
<p class="p2">The <span class="s1">fitnessEffect()</span> callback mechanism is quite flexible and useful, although it has been considerably eclipsed by the modern modern and efficient <span class="s1">fitnessScaling</span> property mentioned above.<span class="Apple-converted-space">  </span>When efficiency is not at a premium, it remains a clear and expressive paradigm for modeling individual-level fitness effects.<span class="Apple-converted-space">  </span>The performance penalty paid is often not large, since these callbacks are called only once per individual per tick, whereas a <span class="s1">mutationEffect()</span> for a type of mutation that is common in the simulation might be called thousands of times per individual per tick (once per mutation of that type possessed by the focal individual).<span class="Apple-converted-space">  </span>The performance penalty typically becomes severe only when the <span class="s1">fitnessEffect()</span> callback needs to perform calculations, once per focal individual, that would vectorize well if performed across a whole vector of individuals.<span class="Apple-converted-space">  </span>In such cases, <span class="s1">fitnessScaling</span> should be used.</p>
<p class="p1"><i>5.13.4<span class="Apple-converted-space">  </span>ITEM: 5. </i><span class="s1"><i>mateChoice()</i></span><i> callbacks</i></p>
<p class="p2">Normally, WF models in SLiM regulate mate choice according to fitness; individuals of higher fitness are more likely to be chosen as mates.<span class="Apple-converted-space">  </span>However, one might wish to simulate more complex mate-choice dynamics such as assortative or disassortative mating, mate search algorithms, and so forth.<span class="Apple-converted-space">  </span>Such dynamics can be handled in WF models with the <span class="s1">mateChoice()</span> callback mechanism.<span class="Apple-converted-space">  </span>(In nonWF models mating is arranged by the script, so there is no need for a callback.)</p>
<p class="p2">A <span class="s1">mateChoice()</span> callback is established in the input file with a syntax very similar to that of <span class="s1">fitnessEffect()</span> callbacks:</p>
<p class="p3">[id] [t1 [: t2]] mateChoice([&lt;subpop-id&gt;]) { ... }</p>
<p class="p7">(In multispecies models, the definition must be preceded by a <span class="s1">species</span> specification as usual.)</p>
<p class="p2">Note that if a subpopulation is given to which the <span class="s1">mateChoice()</span> callback is to apply, the callback is used for all matings that will generate a <i>child</i> in the stated subpopulation (as opposed to all matings of <i>parents</i> in the stated subpopulation); this distinction is important when migration causes children in one subpopulation to be generated by matings of parents in a different subpopulation.</p>
<p class="p2">When a <span class="s1">mateChoice()</span> callback is defined, the first parent in a mating is still chosen proportionally according to fitness (if you wish to influence that choice, you can use a <span class="s1">mutationEffect()</span> or <span class="s1">fitnessEffect()</span> callback).<span class="Apple-converted-space">  </span>In a sexual (rather than hermaphroditic) simulation, this will be the female parent; SLiM does not currently support males as the choosy sex.<span class="Apple-converted-space">  </span>The second parent – the male parent, in a sexual simulation – will then be chosen based upon the results of the <span class="s1">mateChoice()</span> callback.</p>
<p class="p2">More specifically, the callback must return a vector of weights, one for each individual in the subpopulation; SLiM will then choose a parent with probability proportional to weight.<span class="Apple-converted-space">  </span>The <span class="s1">mateChoice()</span> callback could therefore modify or replace the standard fitness-based weights depending upon some other criterion such as assortativeness.<span class="Apple-converted-space">  </span>A singleton object of type <span class="s1">Individual</span> may be returned instead of a weights vector to indicate that that specific individual has been chosen as the mate (beginning in SLiM 2.3); this could also be achieved by returned a vector of weights in which the chosen mate has a non-zero weight and all other weights are zero, but returning the chosen individual directly is much more efficient.<span class="Apple-converted-space">  </span>A zero-length return vector – as generated by <span class="s1">float(0)</span>, for example – indicates that a suitable mate was not found; in that event, a new first parent will be drawn from the subpopulation.<span class="Apple-converted-space">  </span>Finally, if the callback returns <span class="s1">NULL</span>, that signifies that SLiM should use the standard fitness-based weights to choose a mate; the <span class="s1">mateChoice()</span> callback did not wish to alter the standard behavior for the current mating (this is equivalent to returning the unmodified vector of weights, but returning <span class="s1">NULL</span> is much faster since it allows SLiM to drop into an optimized case).<span class="Apple-converted-space">  </span>Apart from the special cases described above – a singleton <span class="s1">Individual</span>, <span class="s1">float(0)</span>, and <span class="s1">NULL</span> – the returned vector of weights must contain the same number of values as the size of the subpopulation, and all weights must be non-negative.<span class="Apple-converted-space">  </span>Note that the vector of weights is not required to sum to <span class="s1">1</span>, however; SLiM will convert relative weights on any scale to probabilities for you.</p>
<p class="p2">If the sum of the returned weights vector is zero, SLiM treats it as meaning the same thing as a return of <span class="s1">float(0)</span> – a suitable mate could not be found, and a new first parent will thus be drawn.<span class="Apple-converted-space">  </span>(This is a change in policy beginning in SLiM 2.3; prior to that, returning a vector of sum zero was considered a runtime error.)<span class="Apple-converted-space">  </span>There is a subtle difference in semantics between this and a return of <span class="s1">float(0)</span>: returning <span class="s1">float(0)</span> immediately short-circuits mate choice for the current first parent, whereas returning a vector of zeros allows further applicable <span class="s1">mateChoice()</span> callbacks to be called, one of which might “rescue” the first parent by returning a non-zero weights vector or an individual.<span class="Apple-converted-space">  </span>In most models this distinction is irrelevant, since chaining <span class="s1">mateChoice()</span> callbacks is uncommon.<span class="Apple-converted-space">  </span>When the choice is otherwise unimportant, returning <span class="s1">float(0)</span> will be handled more quickly by SLiM.</p>
<p class="p2">In addition to the standard SLiM globals, a <span class="s1">mateChoice()</span> callback is supplied with some additional information passed through “pseudo-parameters”:</p>
<p class="p9"><span class="s1">individual</span><span class="Apple-tab-span">	</span>The parent already chosen (the female, in sexual simulations)</p>
<p class="p9"><span class="s1">subpop</span><span class="Apple-tab-span">	</span>The subpopulation into which the offspring will be placed</p>
<p class="p9"><span class="s1">sourceSubpop</span><span class="Apple-tab-span">	</span>The subpopulation from which the parents are being chosen</p>
<p class="p10"><span class="s1">weights</span><span class="Apple-tab-span">	</span>The standard fitness-based weights for all individuals</p>
<p class="p2">If sex is enabled, the <span class="s1">mateChoice()</span> callback must ensure that the appropriate weights are zero and nonzero to guarantee that all eligible mates are male (since the first parent chosen is always female, as explained above).<span class="Apple-converted-space">  </span>In other words, weights for females must be <span class="s1">0</span>.<span class="Apple-converted-space">  </span>The <span class="s1">weights</span> vector given to the callback is guaranteed to satisfy this constraint.<span class="Apple-converted-space">  </span>If sex is not enabled – in a hermaphroditic simulation, in other words – this constraint does not apply.</p>
<p class="p2">For example, a simple <span class="s1">mateChoice()</span> callback might look like this:</p>
<p class="p3">1000:2000 mateChoice(p2) {<br>
<span class="Apple-tab-span">	</span>return weights ^ 2;<br>
}</p>
<p class="p2">This defines a <span class="s1">mateChoice()</span> callback for ticks <span class="s1">1000</span> to <span class="s1">2000</span> for subpopulation <span class="s1">p2</span>.<span class="Apple-converted-space">  </span>The callback simply transforms the standard fitness-based probabilities by squaring them.<span class="Apple-converted-space">  </span>Code like this could represent a situation in which fitness and mate choice proceed normally in one subpopulation (<span class="s1">p1</span>, here, presumably), but are altered by the effects of a social dominance hierarchy or male-male competition in another subpopulation (<span class="s1">p2</span>, here), such that the highest-fitness individuals tend to be chosen as mates more often than their (perhaps survival-based) fitness values would otherwise suggest.<span class="Apple-converted-space">  </span>Note that by basing the returned weights on the <span class="s1">weights</span> vector supplied by SLiM, the requirement that females be given weights of <span class="s1">0</span> is finessed; in other situations, care would need to be taken to ensure that.</p>
<p class="p2">More than one <span class="s1">mateChoice()</span> callback may be defined to operate in the same tick.<span class="Apple-converted-space">  </span>As with Eidos events, multiple callbacks will be called in the order in which they were defined.<span class="Apple-converted-space">  </span>Furthermore, each callback will be given the <span class="s1">weights</span> vector returned by the previous callback – so the value of <span class="s1">weights</span> is not necessarily the default fitness-based weights, in fact, but is the result of all previous <span class="s1">weights()</span> callbacks for the current mate-choice event.<span class="Apple-converted-space">  </span>In this way, the effects of multiple callbacks can “stack”.<span class="Apple-converted-space">  </span>If any <span class="s1">mateChoice()</span> callback returns <span class="s1">float(0)</span>, however – indicating that no eligible mates exist, as described above – then the remainder of the callback chain will be short-circuited and a new first parent will immediately be chosen.</p>
<p class="p2">Note that matings in SLiM do not proceed in random order.<span class="Apple-converted-space">  </span>Offspring are generated for each subpopulation in turn, and within each subpopulation the order of offspring generation is also non-random with respect to both the source subpopulation and the sex of the offspring.<span class="Apple-converted-space">  </span>It is important, therefore, that <span class="s1">mateChoice()</span> callbacks are not in any way biased by the offspring generation order; they should not treat matings early in the process any differently than matings late in the process.<span class="Apple-converted-space">  </span>Any failure to guarantee such invariance could lead to large biases in the simulation outcome.<span class="Apple-converted-space">  </span>In particular, it is usually dangerous to activate or deactivate <span class="s1">mateChoice()</span> callbacks while offspring generation is in progress.</p>
<p class="p2">A wide variety of mate choice algorithms can easily be implemented with <span class="s1">mateChoice()</span> callbacks.<span class="Apple-converted-space">  </span>However, <span class="s1">mateChoice()</span> callbacks can be particularly slow since they are called for every proposed mating, and the vector of mating weights can be large and slow to process.</p>
<p class="p1"><i>5.13.5<span class="Apple-converted-space">  </span>ITEM: 6. </i><span class="s1"><i>modifyChild()</i></span><i> callbacks</i></p>
<p class="p2">Normally, a SLiM simulation defines child generation with its rules regarding selfing versus crossing, recombination, mutation, and so forth.<span class="Apple-converted-space">  </span>However, one might wish to modify these rules in particular circumstances – by preventing particular children from being generated, by modifying the generated children in particular ways, or by generating children oneself.<span class="Apple-converted-space">  </span>All of these dynamics can be handled in SLiM with the <span class="s1">modifyChild()</span> callback mechanism.</p>
<p class="p2">A <span class="s1">modifyChild()</span> callback is established in the input file with a syntax very similar to that of other callbacks:</p>
<p class="p3">[id] [t1 [: t2]] modifyChild([&lt;subpop-id&gt;]) { ... }</p>
<p class="p2">The <span class="s1">modifyChild()</span> callback may optionally be restricted to the children generated to occupy a specified subpopulation.<span class="Apple-converted-space">  </span>(In multispecies models, the definition must be preceded by a <span class="s1">species</span> specification as usual.)</p>
<p class="p2">When a <span class="s1">modifyChild()</span> callback is called, a parent or parents have already been chosen, and a candidate child has already been generated.<span class="Apple-converted-space">  </span>The parent or parents are provided to the callback, as is the generated child.<span class="Apple-converted-space">  </span>The callback may accept the generated child, modify it, substitute completely different genetic information for it, or reject it (causing a new parent or parents to be selected and a new child to be generated, which will again be passed to the callback).</p>
<p class="p2">In addition to the standard SLiM globals, a <span class="s1">modifyChild()</span> callback is supplied with additional information passed through “pseudo-parameters”:</p>
<p class="p11"><span class="s1">child</span><span class="Apple-tab-span">	</span>The generated child (an object of class <span class="s1">Individual</span>)</p>
<p class="p11"><span class="s1">isCloning</span><span class="Apple-tab-span">	</span><span class="s1">T</span> if the child is the result of cloning</p>
<p class="p11"><span class="s1">isSelfing</span><span class="Apple-tab-span">	</span><span class="s1">T</span> if the child is the result of selfing (but see note below)</p>
<p class="p11"><span class="s1">parent1</span><span class="Apple-tab-span">	</span>The first parent (an object of class <span class="s1">Individual</span>)</p>
<p class="p11"><span class="s1">parent2</span><span class="Apple-tab-span">	</span>The second parent (an object of class <span class="s1">Individual</span>)</p>
<p class="p11"><span class="s1">subpop</span><span class="Apple-tab-span">	</span>The subpopulation in which the child will live</p>
<p class="p12"><span class="s1">sourceSubpop</span><span class="Apple-tab-span">	</span>The subpopulation of the parents (<span class="s1">==subpop</span> if not a migration mating)</p>
<p class="p2">These may be used in the <span class="s1">modifyChild()</span> callback to decide upon a course of action.<span class="Apple-converted-space">  </span>The haplosomes of child (available as <span class="s1">child.haplosomes</span>) may be modified by the callback; whatever mutations they contain on exit will be used for the new child.<span class="Apple-converted-space">  </span>Alternatively, they may be left unmodified (to accept the generated child as is).<span class="Apple-converted-space">  </span>The child’s haplosomes may be thought of as the two gametes that will fuse to produce the fertilized egg that results in a new offspring; for a biparental cross involving diploid autosomes, <span class="s1">child.haploidGenome1</span> is the gamete contributed by the first parent (the female, if sex is turned on), and <span class="s1">child.haploidGenome2</span> is the gamete contributed by the second parent (the male, if sex is turned on).<span class="Apple-converted-space">  </span>The <span class="s1">child</span> object itself may also be modified – for example, to set the spatial position of the child.</p>
<p class="p2">Importantly, a <span class="s1">logical</span> singleton return value is required from <span class="s1">modifyChild()</span> callbacks.<span class="Apple-converted-space">  </span>Normally this should be <span class="s1">T</span>, indicating that generation of the child may proceed (with whatever modifications might have been made to the child’s haplosomes).<span class="Apple-converted-space">  </span>A return value of <span class="s1">F</span> indicates that generation of this child should not continue; this will cause new parent(s) to be drawn, a new child to be generated, and a new call to the <span class="s1">modifyChild()</span> callback.<span class="Apple-converted-space">  </span>A <span class="s1">modifyChild()</span> callback that always returns <span class="s1">F</span> can cause SLiM to hang, so be careful that it is guaranteed that your callback has a nonzero probability of returning <span class="s1">T</span> for every state your simulation can reach.</p>
<p class="p2">Note that <span class="s1">isSelfing</span> is <span class="s1">T</span> only when a mating was explicitly set up to be a selfing event by SLiM; an individual may also mate with itself by chance (by drawing itself as a mate) even when SLiM did not explicitly set up a selfing event, which one might term <i>incidental</i> selfing.<span class="Apple-converted-space">  </span>If you need to know whether a mating event was an incidental selfing event, you can compare the parents; self-fertilization will always entail <span class="s1">parent1==parent2</span>, even when <span class="s1">isSelfing</span> is <span class="s1">F</span>.<span class="Apple-converted-space">  </span>Since selfing is enabled only in non-sexual simulations, <span class="s1">isSelfing</span> will always be <span class="s1">F</span> in sexual simulations (and incidental selfing is also impossible in sexual simulations).</p>
<p class="p2">Note that matings in SLiM do not proceed in random order.<span class="Apple-converted-space">  </span>Offspring are generated for each subpopulation in turn, and within each subpopulation the order of offspring generation is also non-random with respect to the source subpopulation, the sex of the offspring, and the reproductive mode (selfing, cloning, or autogamy).<span class="Apple-converted-space">  </span>It is important, therefore, that <span class="s1">modifyChild()</span> callbacks are not in any way biased by the offspring generation order; they should not treat offspring generated early in the process any differently than offspring generated late in the process.<span class="Apple-converted-space">  </span>Similar to <span class="s1">mateChoice()</span> callbacks, any failure to guarantee such invariance could lead to large biases in the simulation outcome.<span class="Apple-converted-space">  </span>In particular, it is usually dangerous to activate or deactivate <span class="s1">modifyChild()</span> callbacks while offspring generation is in progress.<span class="Apple-converted-space">  </span>When SLiM sees that <span class="s1">mateChoice()</span> or <span class="s1">modifyChild()</span> callbacks are defined, it randomizes the order of child generation within each subpopulation, so this issue is mitigated somewhat.<span class="Apple-converted-space">  </span>However, offspring are still generated for each subpopulation in turn.<span class="Apple-converted-space">  </span>Furthermore, in ticks without active callbacks offspring generation order will not be randomized (making the order of parents nonrandom in the next generation), with possible side effects.<span class="Apple-converted-space">  </span>In short, order-dependency issues are possible and must be handled very carefully.</p>
<p class="p2">As with the other callback types, multiple <span class="s1">modifyChild()</span> callbacks may be registered and active.<span class="Apple-converted-space">  </span>In this case, all registered and active callbacks will be called for each child generated, in the order that the callbacks were registered.<span class="Apple-converted-space">  </span>If a <span class="s1">modifyChild()</span> callback returns <span class="s1">F</span>, however, indicating that the child should not be generated, the remaining callbacks in the chain will not be called.</p>
<p class="p2">There are many different ways in which a <span class="s1">modifyChild()</span> callback could be used in a simulation.<span class="Apple-converted-space">  </span>In nonWF models, <span class="s1">modifyChild()</span> callbacks are often unnecessary since each generated child is available to the script in the models’ <span class="s1">reproduction()</span> callback anyway; but they may be used if desired.</p>
<p class="p1"><i>5.13.6<span class="Apple-converted-space">  </span>ITEM: 7. </i><span class="s1"><i>recombination()</i></span><i> callbacks</i></p>
<p class="p2">Typically, a simulation sets up a recombination map at the beginning of the run with <span class="s1">initializeRecombinationRate()</span>, and that map is used for the duration of the run.<span class="Apple-converted-space">  </span>Less commonly, the recombination map is changed dynamically from tick to tick, with <span class="s1">Chromosome</span>’s method <span class="s1">setRecombinationRate()</span>; but still, a single recombination map applies for all individuals of a species in a given tick.<span class="Apple-converted-space">  </span>However, in unusual circumstances a simulation may need to modify the way that recombination works on an individual basis; for this, the <span class="s1">recombination()</span> callback mechanism is provided.<span class="Apple-converted-space">  </span>This can be useful for models involving chromosomal inversions that prevent recombination within a region for some individuals, for example, or for models of the evolution of recombination.</p>
<p class="p2">A <span class="s1">recombination()</span> callback is defined with a syntax much like that of other callbacks:</p>
<p class="p3">[id] [t1 [: t2]] recombination([&lt;subpop-id&gt; [, &lt;chromosome-id&gt;]]) { ... }</p>
<p class="p2">The <span class="s1">recombination()</span> callback will be called during the generation of every gamete during the tick(s) in which it is active.<span class="Apple-converted-space">  </span>It may optionally be restricted to apply only to gametes generated by parents in a specified subpopulation, using the <span class="s1">&lt;subpop-id&gt;</span> specifier.<span class="Apple-converted-space">  </span>In addition, in multi-chromosome models it may optionally be restricted to apply only to a specified chromosome, using the <span class="s1">&lt;chromosome-id&gt;</span> specifier, which may be either the <span class="s1">id</span> or the <span class="s1">symbol</span> of a chromosome defined in the species.<span class="Apple-converted-space">  </span>(In multispecies models, the definition must be preceded by a <span class="s1">species</span> specification as usual.)</p>
<p class="p2">When a <span class="s1">recombination()</span> callback is called, a parent has already been chosen to generate a gamete, and candidate recombination breakpoints for use in recombining the parental haplosomes have been drawn.<span class="Apple-converted-space">  </span>The relevant haplosomes of the focal parent are provided to the callback, as is the focal parent itself (as an <span class="s1">Individual</span> object) and the subpopulation in which it resides.<span class="Apple-converted-space">  </span>Furthermore, the proposed breakpoints are provided to the callback.<span class="Apple-converted-space">  </span>The callback may modify these breakpoints in order to change the breakpoints used, in which case it must return <span class="s1">T</span> to indicate that changes were made, or it may leave the proposed breakpoints unmodified, in which case it must return <span class="s1">F</span>.<span class="Apple-converted-space">  </span>(The behavior of SLiM is undefined if the callback returns the wrong <span class="s1">logical</span> value.)</p>
<p class="p2">In addition to the standard SLiM globals, then, a <span class="s1">recombination()</span> callback is supplied with additional information passed through “pseudo-parameters”:</p>
<p class="p9"><span class="s1">individual</span><span class="Apple-tab-span">	</span>The focal parent that is generating a gamete</p>
<p class="p9"><span class="s1">haplosome1</span><span class="Apple-tab-span">	</span>One haplosome of the focal parent; this is the initial copy strand</p>
<p class="p9"><span class="s1">haplosome2</span><span class="Apple-tab-span">	</span>The other haplosome of the focal parent</p>
<p class="p9"><span class="s1">subpop</span><span class="Apple-tab-span">	</span>The subpopulation to which the focal parent belongs</p>
<p class="p13"><span class="s1">breakpoints</span><span class="Apple-tab-span">	</span>An <span class="s1">integer</span> vector of crossover breakpoints</p>
<p class="p2">These may be used in the <span class="s1">recombination()</span> callback to determine the final recombination breakpoints used by SLiM.<span class="Apple-converted-space">  </span>If values are set into <span class="s1">breakpoints</span>, the new values must be of type <span class="s1">integer</span>.<span class="Apple-converted-space">  </span>If <span class="s1">breakpoints</span> is modified by the callback, <span class="s1">T</span> should be returned, otherwise <span class="s1">F</span> should be returned (this is a speed optimization, so that SLiM does not have to spend time checking for changes when no changes have been made).</p>
<p class="p2">The positions specified in <span class="s1">breakpoints</span> mean that a crossover will occur immediately <i>before</i> the specified base position (between the preceding base and the specified base, in other words).<span class="Apple-converted-space">  </span>The haplosome specified by <span class="s1">haplosome1</span> will be used as the initial copy strand when SLiM executes the recombination; this cannot presently be changed by the callback.<span class="Apple-converted-space">  </span>(Note that <span class="s1">haplosome1</span> and <span class="s1">haplosome2</span> will be haplosomes from <span class="s1">individual</span>, but their order may be swapped, depending on which is the initial copy strand!)</p>
<p class="p2">In this design, the recombination callback does not specify a custom recombination map.<span class="Apple-converted-space">  </span>Instead, the callback can add or remove breakpoints at specific locations.<span class="Apple-converted-space">  </span>To implement a chromosomal inversion, for example, if the parent is heterozygous for the inversion mutation then crossovers within the inversion region are removed by the callback.<span class="Apple-converted-space">  </span>As another example, to implement a model of the evolution of the overall recombination rate, a model could (1) set the global recombination rate to the highest rate attainable in the simulation, (2) for each individual, within the <span class="s1">recombination()</span> callback, calculate the fraction of that maximum rate that the focal individual would experience based upon its genetics, and (3) probabilistically remove proposed crossover points based upon random uniform draws compared to that threshold fraction, thus achieving the individual effective recombination rate desired.<span class="Apple-converted-space">  </span>Other similar treatments could actually vary the effective recombination map, not just the overall rate, by removing proposed crossovers with probabilities that depend upon their position, allowing for the evolution of localized recombination hot-spots and cold-spots.<span class="Apple-converted-space">  </span>Crossovers may also be added, not just removed, by <span class="s1">recombination()</span> callbacks.</p>
<p class="p2">In SLiM 3.3 the recombination model in SLiM was redesigned.<span class="Apple-converted-space">  </span>This required a corresponding redesign of <span class="s1">recombination()</span> callbacks.<span class="Apple-converted-space">  </span>In particular, the <span class="s1">gcStarts</span> and <span class="s1">gcEnds</span> pseudo-parameters to <span class="s1">recombination()</span> callbacks were removed.<span class="Apple-converted-space">  </span>In the present design, the callback receives “crossover breakpoints” information only, in the <span class="s1">breakpoints</span> pseudo-parameter; it receives no information about gene conversion.<span class="Apple-converted-space">  </span>However, <span class="s1">recombination()</span> callbacks can still be used with the “DSB” recombination model; at the point when the callback is called, the pattern of gene conversion tracts will have been simplified down to a vector of crossover breakpoints.<span class="Apple-converted-space">  </span>“Complex” gene conversion tracts, however, involving heteroduplex mismatch repair, are not compatible with <span class="s1">recombination()</span> callbacks, since there is presently no way for them to be specified to the callback.</p>
<p class="p2">Note that the positions in <span class="s1">breakpoints</span> are not, in the general case, guaranteed to be sorted or uniqued; in other words, positions may appear out of order, and the same position may appear more than once.<span class="Apple-converted-space">  </span>After all <span class="s1">recombination()</span> callbacks have completed, the positions from <span class="s1">breakpoints</span> will be sorted, uniqued, and used as the crossover points in generating the prospective gamete haplosome.<span class="Apple-converted-space">  </span>The essential point here is that if the same position occurs more than once, across <span class="s1">breakpoints</span>, the multiple occurrences of the position do not cancel; SLiM does not cross over and then “cross back over” given a pair of identical positions.<span class="Apple-converted-space">  </span>Instead, the multiple occurrences of the position will simply be uniqued down to a single occurrence.</p>
<p class="p2">As with the other callback types, multiple <span class="s1">recombination()</span> callbacks may be registered and active.<span class="Apple-converted-space">  </span>In this case, all registered and active callbacks will be called for each gamete generated, in the order that the callbacks were registered.</p>
<p class="p1"><i>5.13.7<span class="Apple-converted-space">  </span>ITEM: 8. </i><span class="s1"><i>interaction()</i></span><i> callbacks</i></p>
<p class="p2">The <span class="s1">InteractionType</span> class provides various built-in interaction functions that translate from distances to interaction strengths.<span class="Apple-converted-space">  </span>However, it may sometimes be useful to define a custom function for that purpose; for that reason, SLiM allows <span class="s1">interaction()</span> callbacks to be defined that modify the standard interaction strength calculated by <span class="s1">InteractionType</span>.<span class="Apple-converted-space">  </span>In particular, this mechanism allows the strength of interactions to depend upon not only the distance between individuals, but also the genetics and other state of the individuals, the spatial position of the individuals, and other environmental variables.</p>
<p class="p2">An <span class="s1">interaction()</span> callback is called by SLiM when it is determining the strength of the interaction between one individual (the <i>receiver</i> of the interaction) and another individual (the <i>exerter</i> of the interaction).<span class="Apple-converted-space">  </span>This generally occurs when an interaction query is made to <span class="s1">InteractionType</span>, as a side effect of serving that query.<span class="Apple-converted-space">  </span>This means that <span class="s1">interaction()</span> callbacks may be called at a variety of points in the tick cycle, unlike the other callback types in SLiM, which are each called at a specific point.<span class="Apple-converted-space">  </span>If you write an <span class="s1">interaction()</span> callback, you need to take this into account; assuming that the tick cycle is at a particular stage, or even that the tick or cycle is the same as it was when <span class="s1">evaluate()</span> was called, may be dangerous.</p>
<p class="p2">When an interaction strength is needed, the first thing SLiM does is calculate the default interaction strength using the interaction function that has been defined for the <span class="s1">InteractionType</span>.<span class="Apple-converted-space">  </span>If the receiver is the same as the exerter, the interaction strength is always zero; and in spatial simulations if the distance between the receiver and the exerter is greater than the maximum distance set for the <span class="s1">InteractionType</span>, the interaction strength is also always zero.<span class="Apple-converted-space">  </span>In these cases, <span class="s1">interaction()</span> callbacks will not be called, and there is no way to redefine these interaction strengths.</p>
<p class="p2">Otherwise, SLiM will then call <span class="s1">interaction()</span> callbacks that apply to the interaction type and exerter subpopulation for the interaction being evaluated.<span class="Apple-converted-space">  </span>An <span class="s1">interaction()</span> callback is defined with a variation of the syntax used for other callbacks:</p>
<p class="p3">[id] [t1 [: t2]] interaction(&lt;int-type-id&gt; [, &lt;subpop-id&gt;]) { ... }</p>
<p class="p2">For example, if the callback were defined as:</p>
<p class="p3">1000:2000 interaction(i2, p3) { 1.0; }</p>
<p class="p7">then an interaction strength of <span class="s1">1.0</span> would be used for all interactions of interaction type <span class="s1">i2</span>, for exerters in subpopulation <span class="s1">p3</span>, from tick <span class="s1">1000</span> to tick <span class="s1">2000</span>.</p>
<p class="p2">Beginning in SLiM 4, the receiver and exerter may be in different subpopulations from each other – or even, in multispecies models, in different species altogether.<span class="Apple-converted-space">  </span>For the subpopulation id in the <span class="s1">interaction()</span> callback declaration, it does not matter which subpopulation the receiver is in; if the exerter is in <span class="s1">p3</span>, for the above example, then the <span class="s1">interaction()</span> callback will be called regardless of the receiver’s subpopulation (assuming other preconditions are also met, such as the tick range and the interaction type id).<span class="Apple-converted-space">  </span>This means that <span class="s1">interaction()</span> callbacks are not species-specific, unlike other callback types; even if an <span class="s1">interaction()</span> callback is declared to be specific to exerters in <span class="s1">p3</span>, as above, receivers can still be in a different species.<span class="Apple-converted-space">  </span>With no subpopulation id specified, <span class="s1">interaction()</span> callbacks are even more general: the <span class="s1">InteractionType</span> can then be evaluated and queried for receivers and exerters belonging to any species.<span class="Apple-converted-space">  </span>For this reason, in multispecies models <span class="s1">interaction()</span> callbacks must be declared using a species specifier of <span class="s1">species all</span>, unlike all other SLiM callback types; it is not legal to declare an <span class="s1">interaction()</span> callback as species-specific.<span class="Apple-converted-space">  </span>Note that there is no way to declare an <span class="s1">interaction()</span> callback as applying only to receivers in a given subpopulation; if that functionality is desired, you can test <span class="s1">receiver.subpopulation</span> in your callback code and act accordingly.</p>
<p class="p2">In addition to the standard SLiM globals, an <span class="s1">interaction()</span> callback is supplied with some additional information passed through “pseudo-parameters”:</p>
<p class="p8"><span class="s1">distance</span><span class="Apple-tab-span">	</span>The distance from receiver to exerter, in spatial simulations; <span class="s1">NAN</span> otherwise</p>
<p class="p8"><span class="s1">strength</span><span class="Apple-tab-span">	</span>The default interaction strength calculated by the interaction function</p>
<p class="p8"><span class="s1">receiver</span><span class="Apple-tab-span">	</span>The individual receiving the interaction (an object of class <span class="s1">Individual</span>)</p>
<p class="p5"><span class="s1">exerter</span><span class="Apple-tab-span">	</span>The individual exerting the interaction (an object of class <span class="s1">Individual</span>)</p>
<p class="p2">These may be used in the <span class="s1">interaction()</span> callback to compute an interaction strength.<span class="Apple-converted-space">  </span>To simply use the default interaction strength that SLiM would use if a callback had not been defined for interaction type <span class="s1">i1</span>, for example, you could do this:</p>
<p class="p3">interaction(i1) {<br>
<span class="Apple-tab-span">	</span>return strength;<br>
}</p>
<p class="p2">Usually an <span class="s1">interaction()</span> callback will modify that default strength based upon factors such as the genetics of the receiver and/or the exerter, the spatial positions of the two individuals, or some other simulation state.<span class="Apple-converted-space">  </span>Any finite <span class="s1">float</span> value greater than or equal to <span class="s1">0.0</span> may be returned.<span class="Apple-converted-space">  </span>The value returned will be not be cached by SLiM; if the interaction strength between the same two individuals is needed again later, the <span class="s1">interaction()</span> callback will be called again (something to keep in mind if the interaction strength includes a stochastic component).<span class="Apple-converted-space">  </span>Note that the provided <span class="s1">distance</span> and <span class="s1">strength</span> values are based upon the spatial positions of the exerter and receiver when <span class="s1">evaluate()</span> was called, not their current spatial positions, if they have moved since the interaction was evaluated.</p>
<p class="p2">More than one <span class="s1">interaction()</span> callback may be defined to operate in the same tick.<span class="Apple-converted-space">  </span>As with other callbacks, multiple callbacks will be called in the order in which they were defined in the input file.<span class="Apple-converted-space">  </span>Furthermore, each callback will be given the <span class="s1">strength</span> value returned by the previous callback – so the value of <span class="s1">strength</span> is not necessarily the default value, in fact, but is the result of all previous <span class="s1">interaction()</span> callbacks for the interaction in question.<span class="Apple-converted-space">  </span>In this way, the effects of multiple callbacks can “stack”.</p>
<p class="p2">The <span class="s1">interaction()</span> callback mechanism is extremely powerful and flexible, allowing any sort of user-defined interactions whatsoever to be queried dynamically using the methods of <span class="s1">InteractionType</span>.<span class="Apple-converted-space">  </span>However, in the general case a simulation may call for the evaluation of the interaction strength between each individual and every other individual, making the computation of the full interaction network an O(<i>N</i><span class="s7"><sup>2</sup></span>) problem.<span class="Apple-converted-space">  </span>Since <span class="s1">interaction()</span> callbacks may be called for each of those <i>N</i><span class="s7"><sup>2</sup></span> interaction evaluations, they can slow down a simulation considerably, so it is recommended that they be used sparingly.<span class="Apple-converted-space">  </span>This is the reason that the various interaction functions of <span class="s1">InteractionType</span> were provided; when an interaction does not depend upon individual state, the intention is to avoid the necessity of an <span class="s1">interaction()</span> callback altogether.<span class="Apple-converted-space">  </span>Furthermore, constraining the number of cases in which interaction strengths need to be calculated – using a short maximum interaction distance, querying the nearest neighbors of the focal individual rather than querying all possible interactions with that individual, and specifying the reciprocality and sex segregation of the <span class="s1">InteractionType</span>, for example – may greatly decrease the computational overhead of interaction evaluation.</p>
<p class="p1"><i>5.13.8<span class="Apple-converted-space">  </span>ITEM: 9. </i><span class="s1"><i>reproduction()</i></span><i> callbacks</i></p>
<p class="p2">In WF models (the default model type in SLiM), the SLiM core manages the reproduction of individuals in each tick.<span class="Apple-converted-space">  </span>In nonWF models, however, reproduction is managed by the model script, in <span class="s1">reproduction()</span> callbacks.<span class="Apple-converted-space">  </span>These callbacks may only be defined in nonWF models.</p>
<p class="p2">A <span class="s1">reproduction()</span> callback is defined with a syntax much like that of other callbacks:</p>
<p class="p3">[id] [t1 [: t2]] reproduction([&lt;subpop-id&gt; [, &lt;sex&gt;]]) { ... }</p>
<p class="p2">The <span class="s1">reproduction()</span> callback will be called once for each individual during the tick(s) in which it is active.<span class="Apple-converted-space">  </span>It may optionally be restricted to apply only to individuals in a specified subpopulation, using the <span class="s1">&lt;subpop-id&gt;</span> specifier; this may be a subpopulation specifier such as <span class="s1">p1</span>, or <span class="s1">NULL</span> indicating no restriction.<span class="Apple-converted-space">  </span>It may also optionally be restricted to apply only to individuals of a specified sex (in sexual models), using the <span class="s1">&lt;sex&gt;</span> specifier; this may be <span class="s1">"M"</span> or <span class="s1">"F"</span>, or <span class="s1">NULL</span> indicating no restriction.<span class="Apple-converted-space">  </span>(In multispecies models, the definition must be preceded by a <span class="s1">species</span> specification as usual.)</p>
<p class="p2">When a <span class="s1">reproduction()</span> callback is called, SLiM’s expectation is that the callback will trigger the reproduction of a focal individual by making method calls to add new offspring individuals.<span class="Apple-converted-space">  </span>Typically the offspring added are the offspring of the focal individual, and typically they are added to the subpopulation to which the focal individual belongs, but neither of these is required; a <span class="s1">reproduction()</span> callback may add offspring generated by any parent(s), to any subpopulation in the focal species.<span class="Apple-converted-space">  </span>The focal individual is provided to the callback (as an <span class="s1">Individual</span> object), as is the subpopulation in which it resides.</p>
<p class="p2">A common alternative pattern is for a <span class="s1">reproduction()</span> callback to ignore the focal individual and generate all of the offspring for a species for the current tick, from all parents.<span class="Apple-converted-space">  </span>The callback then sets <span class="s1">self.active</span> to <span class="s1">0</span>, preventing itself from being called again in the current tick; this callback design therefore executes once per tick.<span class="Apple-converted-space">  </span>This can be useful if individuals influence each other’s offspring generation (as in a monogamous-mating model, for example); it can also simply be more efficient when producing offspring in bulk.</p>
<p class="p2">In addition to the usual SLiM globals, then, a <span class="s1">reproduction()</span> callback is supplied with additional information passed through global variables:</p>
<p class="p9"><span class="s1">individual</span><span class="Apple-tab-span">	</span>The focal individual that is expected to reproduce</p>
<p class="p13"><span class="s1">subpop</span><span class="Apple-tab-span">	</span>The subpopulation to which the focal individual belongs</p>
<p class="p2">At present, the return value from <span class="s1">reproduction()</span> callbacks is not used, and must be <span class="s1">void</span> (i.e., a value may not be returned).<span class="Apple-converted-space">  </span>It is possible that other return values will be defined in future.</p>
<p class="p2">It is possible, of course, to do actions unrelated to reproduction inside <span class="s1">reproduction()</span> callbacks, but it is not recommended.<span class="Apple-converted-space">  </span>The <span class="s1">first()</span> event phase of the current tick provides an opportunity for actions immediately before reproduction, and the <span class="s1">early()</span> event phase of the current tick provides an opportunity for actions immediately after reproduction, so only actions that are intertwined with reproduction itself should occur in <span class="s1">reproduction()</span> callbacks.<span class="Apple-converted-space">  </span>Besides providing conceptual clarity, following this design principle will also decrease the probability of bugs, since actions that are unrelated to reproduction should usually not influence or be influenced by the dynamics of reproduction.</p>
<p class="p2">If the <span class="s1">randomizeCallbacks</span> parameter to <span class="s1">initializeSLiMOptions()</span> is <span class="s1">T</span> (the default), the order in which individuals are given an opportunity to reproduce with a call to <span class="s1">reproduction()</span> callbacks will be randomized within each subpopulation.<span class="Apple-converted-space">  </span>This partially mitigates order-dependency issues, although such issues can still arise whenever the effects of a <span class="s1">reproduction()</span> callback are not independent.<span class="Apple-converted-space">  </span>If <span class="s1">randomizeCallbacks</span> is <span class="s1">F</span>, individuals will be given their opportunity to reproduce in sequential order within each subpopulation, greatly increasing the risk of order-dependency problems.</p>
<p class="p2">As with the other callback types, multiple <span class="s1">reproduction()</span> callbacks may be registered and active.<span class="Apple-converted-space">  </span>In this case, all registered and active callbacks will be called for each individual, in the order that the callbacks were registered.</p>
<p class="p1"><i>5.13.9<span class="Apple-converted-space">  </span>ITEM: 10. </i><span class="s1"><i>mutation()</i></span><i> callbacks</i></p>
<p class="p2">SLiM auto-generates new mutations according to the current mutation rate (or rate map) and the genetic structure defined by genomic elements, their genomic element types, the mutation types those genomic element types draw from, and the distribution of fitness effects defined by those mutation types.<span class="Apple-converted-space">  </span>In nucleotide-based models, the nucleotide sequence and the mutation matrix also play a role in determining both the rate of mutation and the nucleotide mutated to.<span class="Apple-converted-space">  </span>In some models it can be desirable to modify these dynamics in some way – altering the selection coefficients of new mutations in some way, changing the mutation type used, dictating the nucleotide to be used, replacing the proposed mutation with a pre-existing mutation at the same position, or even suppressing the proposed mutation altogether.<span class="Apple-converted-space">  </span>To achieve this, one may define a <span class="s1">mutation()</span> callback.</p>
<p class="p2">A <span class="s1">mutation()</span> callback is defined as:</p>
<p class="p3">[id] [t1 [: t2]] mutation([&lt;mut-type-id&gt; [, &lt;subpop-id&gt;]]) { ... }</p>
<p class="p2">The <span class="s1">mutation()</span> callback will be called once for each new auto-generated mutation during the tick(s) in which the callback is active.<span class="Apple-converted-space">  </span>It may optionally be restricted to apply only to mutations of a particular mutation type, using the <span class="s1">&lt;mut-type-id&gt;</span> specifier; this may be a mutation type specifier such as <span class="s1">m1</span>, or <span class="s1">NULL</span> indicating no restriction.<span class="Apple-converted-space">  </span>It may also optionally be restricted to individuals generated by a specified subpopulation (usually – see below for discussion), using the <span class="s1">&lt;subpop-id&gt;</span> specifier; this should be a subpopulation specifier such as <span class="s1">p1</span>.<span class="Apple-converted-space">  </span>(In multispecies models, the definition must be preceded by a <span class="s1">species</span> specification as usual.)</p>
<p class="p2">When a <span class="s1">mutation()</span> callback is called, a focal mutation (provided to the callback as an object of type <span class="s1">Mutation</span>) has just been created by SLiM, referencing a particular position in a parental haplosome (also provided, as an object of type <span class="s1">Haplosome</span>).<span class="Apple-converted-space">  </span>The mutation will not be added to that parental haplosome; rather, the parental haplosome is being copied, during reproduction, to make a gamete or an offspring haplosome, and the mutation is, conceptually, a copying error made during that process.<span class="Apple-converted-space">  </span>It will be added to the offspring haplosome that is the end result of the copying process (which may also involve recombination with another haplosome).<span class="Apple-converted-space">  </span>At the point that the <span class="s1">mutation()</span> callback is called, the offspring haplosome is not yet created, however, and so it cannot be accessed from within the <span class="s1">mutation()</span> callback; the <span class="s1">mutation()</span> callback can affect only the mutation itself, not the haplosome to which the mutation will be added.</p>
<p class="p2">In addition to the standard SLiM globals, then, a <span class="s1">mutation()</span> callback is supplied with additional information passed through global variables:</p>
<p class="p9"><span class="s1">mut</span><span class="Apple-tab-span">	</span>The focal mutation that is being modified or reviewed</p>
<p class="p9"><span class="s1">haplosome</span><span class="Apple-tab-span">	</span>The parental haplosome that is being copied</p>
<p class="p9"><span class="s1">element</span><span class="Apple-tab-span">	</span>The genomic element that controls the mutation site</p>
<p class="p9"><span class="s1">originalNuc</span><span class="Apple-tab-span">	</span>The nucleotide (<span class="s1">0</span>/<span class="s1">1</span>/<span class="s1">2</span>/<span class="s1">3</span> for <span class="s1">A</span>/<span class="s1">C</span>/<span class="s1">G</span>/<span class="s1">T</span>) originally at the mutating position</p>
<p class="p9"><span class="s1">parent</span><span class="Apple-tab-span">	</span>The parent which is generating the offspring haplosome</p>
<p class="p13"><span class="s1">subpop</span><span class="Apple-tab-span">	</span>The subpopulation to which the parent belongs</p>
<p class="p2">The <span class="s1">mutation()</span> callback has three possible returns: <span class="s1">T</span>, <span class="s1">F</span>, or (beginning in SLiM 3.5) a singleton object of type <span class="s1">Mutation</span>.<span class="Apple-converted-space">  </span>A return of <span class="s1">T</span> indicates that the proposed mutation should be used in generating the offspring haplosome (perhaps with modifications made by the callback).<span class="Apple-converted-space">  </span>Conversely, a return of <span class="s1">F</span> indicates that the proposed mutation should be suppressed.<span class="Apple-converted-space">  </span>If a proposed mutation is suppressed, SLiM will not try again; one fewer mutations will be generated during reproduction than would otherwise have been true.<span class="Apple-converted-space">  </span>Returning <span class="s1">F</span> will therefore mean that the realized mutation rate in the model will be lower than the expected mutation rate.<span class="Apple-converted-space">  </span>Finally, a return of an object of type <span class="s1">Mutation</span> replaces the proposed mutation (<span class="s1">mut</span>) with the mutation returned; the offspring haplosomes being generated will contain the returned mutation.<span class="Apple-converted-space">  </span>The position of the returned mutation must match that of the proposed mutation.<span class="Apple-converted-space">  </span>This provides a mechanism for a <span class="s1">mutation()</span> callback to make SLiM re-use existing mutations instead of generating new mutations, which can be useful.</p>
<p class="p2">The callback may perform a variety of actions related to the generated mutation.<span class="Apple-converted-space">  </span>The selection coefficient of the mutation can be changed with <span class="s1">setSelectionCoefficient()</span>, and the mutation type of the mutation can be changed with <span class="s1">setMutationType()</span>; the <span class="s1">drawSelectionCoefficient()</span> method of <span class="s1">MutationType</span> may also be useful here.<span class="Apple-converted-space">  </span>A <span class="s1">tag</span> property value may be set for the mutation, and named values may be attached to the mutation with <span class="s1">setValue()</span>.<span class="Apple-converted-space">  </span>In nucleotide-based models, the <span class="s1">nucleotide</span> (or <span class="s1">nucleotideValue</span>) property of the mutation may also be changed; note that the original nucleotide at the focal position in the parental haplosome is provided through <span class="s1">originalNuc</span> (it could be retrieved with <span class="s1">haplosome.nucleotides()</span>, but SLiM already has it at hand anyway).<span class="Apple-converted-space">  </span>All of these modifications to the new mutation may be based upon the state of the parent, including its genetic state, or upon any other model state.</p>
<p class="p2">It is possible, of course, to do actions unrelated to mutation inside <span class="s1">mutation()</span> callbacks, but it is not recommended; <span class="s1">first()</span>, <span class="s1">early()</span>, and <span class="s1">late()</span> events should be used for general-purpose scripting.<span class="Apple-converted-space">  </span>Besides providing conceptual clarity, following this design principle will also decrease the probability of bugs, since actions that are unrelated to mutation should not influence or be influenced by the dynamics of mutation.</p>
<p class="p2">The proposed mutation will not appear in the <span class="s1">sim.mutations</span> vector of segregating mutations until it has been added to a haplosome; it will therefore not be visible in that vector within its own <span class="s1">mutation()</span> callback invocation, and indeed, may not be visible in subsequent callbacks during the reproduction tick cycle stage until such time as the offspring individual being generated has been completed.<span class="Apple-converted-space">  </span>If that offspring is ultimately rejected, in particular by a <span class="s1">modifyChild()</span> callback, the proposed mutation may not be used by SLiM at all.<span class="Apple-converted-space">  </span>It may therefore be unwise to assume, in a <span class="s1">mutation()</span> callback, that the focal mutation will ultimately be added to the simulation, depending upon the rest of the model’s script.</p>
<p class="p2">There is one subtlety to be mentioned here, having to do with subpopulations.<span class="Apple-converted-space">  </span>The <span class="s1">subpop</span> pseudo-parameter discussed above is always the subpopulation of the parent which possesses the haplosome that is being copied and is mutating; there is no ambiguity about that whatsoever.<span class="Apple-converted-space">  </span>The <span class="s1">&lt;subpop-id&gt;</span> specified in the <span class="s1">mutation()</span> callback declaration, however, is a bit more subtle; above it was said that it restricts the callback “to individuals generated by a specified subpopulation”, and that is usually true but requires some explanation.<span class="Apple-converted-space">  </span>In WF models, recall that migrants are generated in a source subpopulation and placed in a target subpopulation, as a model of juvenile migration; in that context, the <span class="s1">&lt;subpop-id&gt;</span> specifies the <i>source</i> subpopulation to which the <span class="s1">mutation()</span> callback will be restricted.<span class="Apple-converted-space">  </span>In nonWF models, offspring are generated by the <span class="s1">add...()</span> family of <span class="s1">Subpopulation</span> methods, which can cross individuals from two different subpopulations and place the result in a third target subpopulation; in that context, in general, the <span class="s1">&lt;subpop-id&gt;</span> specifies the source subpopulation that is generating the particular <i>gamete</i> that is sustaining a mutation during its production.<span class="Apple-converted-space">  </span>The exception to this rule is <span class="s1">addRecombinant()</span> and <span class="s1">addMultiRecombinant()</span>; since there are four different source subpopulations potentially in play there per mutation, it was deemed simpler in that case for the <span class="s1">&lt;subpop-id&gt;</span> to specify the <i>target</i> subpopulation to which the <span class="s1">mutation()</span> callback will be restricted.<span class="Apple-converted-space">  </span>If restriction to the source subpopulation is needed with <span class="s1">addRecombinant()</span> or <span class="s1">addMultiRecombinant()</span>, the <span class="s1">subpop</span> pseudo-parameter may be consulted rather than using <span class="s1">&lt;subpop-id&gt;</span>.</p>
<p class="p2">Note that <span class="s1">mutation()</span> callbacks are only called for mutations that are auto-generated by SLiM, as a consequence of the mutation rate and the genetic structure defined.<span class="Apple-converted-space">  </span>Mutations that are created in script, using <span class="s1">addNewMutation()</span> or <span class="s1">addNewDrawnMutation()</span>, will not trigger <span class="s1">mutation()</span> callbacks; but of course the script may modify or tailor such added mutations in whatever way is desired, so there is no need for callbacks in that situation.</p>
<p class="p2">As with the other callback types, multiple <span class="s1">mutation()</span> callbacks may be registered and active.<span class="Apple-converted-space">  </span>In this case, all registered and active callbacks will be called for each generated mutation to which they apply, in the order that the callbacks were registered.</p>
<p class="p1"><i>5.13.10<span class="Apple-converted-space">  </span>ITEM: 11. </i><span class="s1"><i>survival()</i></span><i> callbacks</i></p>
<p class="p2">In nonWF models, a selection phase in the tick cycle results in mortality; individuals survive or die based upon their fitness.<span class="Apple-converted-space">  </span>In most cases this standard behavior is sufficient; but occasionally it can be useful to observe the survival decisions SLiM makes (to log out information about dying individuals, for example), to modify those decisions (influencing which individuals live and which die, perhaps based upon factors other than genetics), or even to short-circuit mortality completely (moving dead individuals into a “cold storage” subpopulation for later use, perhaps).<span class="Apple-converted-space">  </span>To accomplish such goals, one can the <span class="s1">survival()</span> callback mechanism to override SLiM’s default behavior.<span class="Apple-converted-space">  </span>Note that in WF models, since they always model non-overlapping generations, the entire parental generation dies in each tick regardless of fitness; <span class="s1">survival()</span> callbacks therefore apply only to nonWF models.</p>
<p class="p2">A <span class="s1">survival()</span> callback is defined with a syntax much like that of other callbacks:</p>
<p class="p3">[id] [t1 [: t2]] survival([&lt;subpop-id&gt;]) { ... }</p>
<p class="p2">The <span class="s1">survival()</span> callback will be called during the selection phase of the tick cycle of nonWF models, during the tick(s) in which it is active.<span class="Apple-converted-space">  </span>By default it will be called once per individual in the entire population (whether slated for survival or not); it may optionally be restricted to apply only to individuals in a specified subpopulation, using the <span class="s1">&lt;subpop-id&gt;</span> specifier.<span class="Apple-converted-space">  </span>(In multispecies models, the definition must be preceded by a <span class="s1">species</span> specification as usual.)</p>
<p class="p2">When a <span class="s1">survival()</span> callback is called, a focal individual has already been evaluated by SLiM regarding its survival; a final fitness value for the individual has been calculated, and a random uniform draw in <span class="s1">[0,1]</span> has been generated that determines whether the individual is to survive (a draw less than the individual’s fitness) or die (a draw greater than or equal to the individual’s fitness).<span class="Apple-converted-space">  </span>The focal individual is provided to the callback, as is the subpopulation in which it resides.<span class="Apple-converted-space">  </span>Furthermore, the preliminary decision (whether the focal individual will survive or not), the focal individual’s fitness, and the random draw made by SLiM to determine survival are also provided to the callback.<span class="Apple-converted-space">  </span>The callback may return <span class="s1">NULL</span> to accept SLiM’s decision, or may return <span class="s1">T</span> to indicate that the individual should survive, or <span class="s1">F</span> to indicate that it should die, regardless of its fitness and the random deviate drawn.<span class="Apple-converted-space">  </span>The callback may also return a singleton <span class="s1">Subpopulation</span> object to indicate the individual should remain alive but should be moved to that subpopulation (note that calling <span class="s1">takeMigrants()</span> during the survival phase is illegal, because SLiM is busy modifying the population’s internal state).</p>
<p class="p2">In addition to the standard SLiM globals, then, a <span class="s1">survival()</span> callback is supplied with additional information passed through “pseudo-parameters”:</p>
<p class="p9"><span class="s1">individual</span><span class="Apple-tab-span">	</span>The focal individual that will live or die</p>
<p class="p9"><span class="s1">subpop</span><span class="Apple-tab-span">	</span>The subpopulation to which the focal individual belongs</p>
<p class="p9"><span class="s1">surviving</span><span class="Apple-tab-span">	</span>A <span class="s1">logical</span> value indicating SLiM’s preliminary decision (<span class="s1">T</span> == survival)</p>
<p class="p9"><span class="s1">fitness</span><span class="Apple-tab-span">	</span>The focal individual’s fitness</p>
<p class="p13"><span class="s1">draw</span><span class="Apple-tab-span">	</span>SLiM’s random uniform deviate, which determined the preliminary decision</p>
<p class="p2">These may be used in the <span class="s1">survival()</span> callback to determine the final decision.</p>
<p class="p2">While <span class="s1">survival()</span> callbacks are still being called, no decisions are put into effect; no individuals actually die, and none are moved to a new <span class="s1">Subpopulation</span> if that was requested.<span class="Apple-converted-space">  </span>In effect, SLiM pre-plans the fate of every individual completely without modifying the model state at all.<span class="Apple-converted-space">  </span>After all <span class="s1">survival()</span> callbacks have completed for every individual, the planned fates for every individual will then be executed, without any opportunity for further intervention through callbacks.<span class="Apple-converted-space">  </span>It is therefore legal to inspect subpopulations and individuals inside a <span class="s1">survival()</span> callback, but it should be understood that previously made decisions about the fates of other individuals will not yet have any visible effect.<span class="Apple-converted-space">  </span>It is generally a good idea for the decisions rendered by <span class="s1">survival()</span> callbacks to be independent anyway, to avoid biases due to order-dependency.<span class="Apple-converted-space">  </span>If the <span class="s1">randomizeCallbacks</span> parameter to <span class="s1">initializeSLiMOptions()</span> is <span class="s1">T</span> (the default), the order in which <span class="s1">survival()</span> callbacks are called on individuals will be randomized within each subpopulation; nevertheless, order-dependency issues can occur if callback effects are not independent.<span class="Apple-converted-space">  </span>If <span class="s1">randomizeCallbacks</span> is <span class="s1">F</span>, the order in which individuals are evaluated within each subpopulation is not guaranteed to be random, and order-dependency problems are thus even more likely.</p>
<p class="p2">It is worth noting that if <span class="s1">survival()</span> callbacks are used, “fitness” in the model is then no longer really fitness; the model is making its own decisions about which individuals live and die, and those decisions are the true determinant of fitness in the biological sense.<span class="Apple-converted-space">  </span>A <span class="s1">survival()</span> callback that makes its own decisions regarding survival with no regard for SLiM’s calculated fitness values can completely alter the pattern of selection in a population, rendering all of SLiM’s fitness machinery – selection and dominance coefficients, <span class="s1">fitnessScaling</span> values, etc. – completely irrelevant.<span class="Apple-converted-space">  </span>To avoid highly counterintuitive and confusing effects, it is thus generally a good idea to use of <span class="s1">survival()</span> callbacks only when it is strictly necessary to achieve a desired outcome.</p>
<p class="p2">As with the other callback types, multiple <span class="s1">survival()</span> callbacks may be registered and active.<span class="Apple-converted-space">  </span>In this case, all registered and active callbacks will be called for each individual evaluated, in the order that the callbacks were registered.</p>
<p class="p14"><span class="s8"></span><br></p>
</body>
</html>
